*** lib/Analysis/TypeBasedAliasAnalysis.cpp.old	2020-01-02 17:46:09.047102194 -0500
--- lib/Analysis/TypeBasedAliasAnalysis.cpp	2020-01-02 17:52:33.489117324 -0500
***************
*** 571,575 ****
    for (unsigned I = 0, E = BaseType.getNumFields(); I != E; ++I) {
      TBAAStructTypeNode T = BaseType.getFieldType(I);
!     if (T == FieldType || hasField(T, FieldType))
        return true;
    }
--- 571,576 ----
    for (unsigned I = 0, E = BaseType.getNumFields(); I != E; ++I) {
      TBAAStructTypeNode T = BaseType.getFieldType(I);
!     if (matchAccessTags(createAccessTag(T.getNode()),
! 			createAccessTag(FieldType.getNode())))
        return true;
    }
