*** lib/Target/NVPTX/NVPTXTargetMachine.cpp.old	2020-03-23 11:01:02.000000000 -0400
--- lib/Target/NVPTX/NVPTXTargetMachine.cpp	2020-10-01 11:51:23.485703369 -0400
***************
*** 100,104 ****
      Ret += "-p3:32:32-p4:32:32-p5:32:32";
  
!   Ret += "-i64:64-i128:128-v16:16-v32:32-n16:32:64";
  
    return Ret;
--- 100,104 ----
      Ret += "-p3:32:32-p4:32:32-p5:32:32";
  
!   Ret += "-i64:64-i128:128-v16:16-v32:32-n16:32:64-S64";
  
    return Ret;
